src/userdev/java/net/minecraftforge/gradle/userdev/DependencyManagementExtension.java

```diff
@@ -68,7 +70,12 @@ public Dependency deobf(Provider<MinimalExternalModuleDependency> provider) {
         MinimalExternalModuleDependency dependency = provider.get();
         project.getConfigurations().getByName(UserDevPlugin.OBF).getDependencies().add(dependency);
 
-        return remapper.remap(dependency);
+        ModuleIdentifier module = dependency.getModule();
+        if(dependency instanceof DependencyVariant) {
+            DependencyVariant variant = (DependencyVariant) dependency;
+            return deobf(String.format("%s:%s:%s:%s", module.getGroup(), module.getName(), dependency.getVersion(), variant.getClassifier()));
+        }
+        return deobf(String.format("%s:%s:%s", module.getGroup(), module.getName(), dependency.getVersion()));
```

Comment:
This doesn't seem right, it causes the dependency to be added twice, once using the object passed in, and the second as a string.

Reply:
How and where? I tested it, it works fine and there's no duplicates, at least not for me..

Reply:
A dependency passed _into_ deobf as a provider is not added to the dep graph in any way yet. `MinimalExternalModuleDependency` is used as a means to define type safe version catalog entries.

Reply:
Yes, but you add the dependency to the graph on line 71, then pass in a string variant to deobf which in turn adds he string variant on line 64.
The issue is that the version will not be updated to include the `_mapped_` identifier that the remapper adds to let the deobf repo know to provide the dep.

You're also depending on internal gradle API using DependencyVaraint. Is there a public version of that API?
I couldn't find one, perhaps it would be better to add a secondary deobf function that took in a `Action<? super ExternalModuleDependencyVariantSpec> variantSpec` and passed it to the `variantOf` function ourselves. this making the syntax potentially `fg.deobf(libs.dep) { classifier('api') }`

Simplest solution I think would be to just remove line 71.

Is there a way to get a guaranteed mutable version of MinimalExternalModuleDependency? Specifically we need to modify the version constraints to include the mapped name.

Perhaps this is time for someone to figure out how to properly use the Variant/Attribute system gradle introduced after all of this was written to encode the required mapping version as an attribute. And then have the Deobf repo provide an artifact with that attribute. I have not looked into the system in years so I am unsure if this is possible.

Reply:
> Perhaps this is time for someone to figure out how to properly use the Variant/Attribute system gradle introduced after all of this was written to encode the required mapping version as an attribute. And then have the Deobf repo provide an artifact with that attribute. I have not looked into the system in years so I am unsure if this is possible.

I'd like to consider myself fairly well versed with Gradle from a plugin perspective, but i'd like to see a better way too and maybe improve my skills. I'll look into it.

> Is there a way to get a guaranteed mutable version of MinimalExternalModuleDependency?

Not any that i'm aware of right now sadly, that's why i've attempted to do it the way i did.

> You're also depending on internal gradle API using DependencyVaraint. Is there a public version of that API?
I couldn't find one, perhaps it would be better to add a secondary deobf function that took in a Action<? super ExternalModuleDependencyVariantSpec> variantSpec and passed it to the variantOf function ourselves. this making the syntax potentially fg.deobf(libs.dep) { classifier('api') }

This sounds really good! ^^ I will implement these changes in a moment.

Reply:
Feel free to join the discord and we can shat about some ideas there. But ya, I personally hate Gradle with a burning passion because they don't know how to make things simple, and don't know how to keep things stable. But if you wanna poke at things, I have a few ideas and it wouldn't hurt to try.